ostd/src/mm/page_table/node/entry.rs

```diff
@@ -94,14 +95,51 @@ impl<'a, E: PageTableEntryTrait, C: PagingConstsTrait> Entry<'a, E, C> {
             *self.node.nr_children_mut() -= 1;
         }
 
+        let new_pte = new_child.into_pte();
+
         // SAFETY:
         //  1. The index is within the bounds.
         //  2. The new PTE is compatible with the page table node, as asserted above.
-        unsafe { self.node.write_pte(self.idx, new_child.into_pte()) };
+        unsafe { self.node.write_pte(self.idx, new_pte) };
+
+        self.pte = new_pte;
 
         old_child
     }
 
+    /// Allocates a new child page table node and replaces the entry with it.
+    ///
+    /// If the old entry is not none, the operation will fail and return `None`.
+    /// Otherwise, the lock guard of the new child page table node is returned.
+    pub(in crate::mm::page_table) fn alloc_if_none(
+        &mut self,
+        new_pt_is_tracked: MapTrackingStatus,
+    ) -> Option<PageTableGuard<'pt, E, C>> {
+        if !self.is_none() {
+            return None;
+        }
+
+        let level = self.node.level();
+        let new_page = PageTableNode::<E, C>::alloc(level - 1, new_pt_is_tracked);
+
+        let guard_addr = new_page.lock().into_raw_paddr();
+
+        // SAFETY:
+        //  1. The index is within the bounds.
+        //  2. The new PTE is compatible with the page table node.
+        unsafe {
+            self.node
+                .write_pte(self.idx, Child::PageTable(new_page).into_pte())
+        };
+
+        *self.node.nr_children_mut() += 1;
+
+        // SAFETY: The resulting guard lifetime (`'a`) is no shorter than the
+        // lifetime of the current entry (`'a`), because we store the allocated
+        // page table in the current node.
```

Comment:
I don't think the safety reasoning is right here:

```rust
let allocated = entry.alloc_if_none();
drop(entry.replace(Child::None));
// Can we use `allocated` here?
```


Even if the `drop()` goes through the RCU (we can use something like `RcuDrop` in `Child`, as I said above), the above code is still problematic because the current implementation does not seem to guarantee anywhere that `'pt` will not outlive the RCU grace period.

I think we should have:
```rust
    pub(super) fn lock(&'rcu self, &'rcu PreemptionGuard) -> PageTableGuard<'rcu, E, C>;
```

So here the safety comments should be updated to:
```rust
        // SAFETY: The page table won't be dropped before the RCU grace period
        // ends, so it outlives `'rcu`.
```